components/places/src/storage.rs

```diff
@@ -263,21 +270,314 @@ fn add_visit(db: &impl ConnExt,
+        )?;
+        Ok(())
+    }
+
+    pub fn apply_synced_deletion(db: &Connection, guid: &SyncGuid) -> Result<()> {
+        db.execute_named_cached(
+            "DELETE FROM moz_places WHERE guid = :guid",
+            &[(":guid", guid)])?;
+        // This might have created a local tombstone for the item, which we
+        // don't want.
+        db.execute_named_cached(
+            "DELETE FROM moz_places_tombstones WHERE guid = :guid",
+            &[(":guid", guid)])?;
+        Ok(())
+    }
+
+    #[derive(Debug)]
+    pub enum OutgoingInfo {
+        Record(HistoryRecord),
+        Tombstone,
+    }
+
+    pub fn fetch_outgoing(db: &impl ConnExt,
+                          max_places: usize,
+                          max_visits: usize) -> Result<HashMap<SyncGuid, OutgoingInfo>> {
+        // Note that we want *all* "new" regardless of change counter,
+        // so that we do the right thing after a "reset".
+        let mut stmt = db.conn().prepare(&format!("
+            SELECT guid, url, id, title, hidden, typed, frecency,
+                visit_count_local, visit_count_remote,
+                last_visit_date_local, last_visit_date_remote,
+                sync_status, sync_change_counter
+            FROM moz_places
+            WHERE (sync_change_counter > 0 OR sync_status != {})
+            ORDER BY frecency DESC
+            LIMIT :max_places", (SyncStatus::Normal as u8)))?;
+        let mut visits = db.conn().prepare("
+            SELECT visit_date as date, visit_type as transition
+            FROM moz_historyvisits
+            WHERE place_id = :place_id
+            ORDER BY visit_date DESC
+            LIMIT :max_visits")?;
+        // tombstones
+        let mut tombstones_stmt = db.conn().prepare("
+            SELECT guid FROM moz_places_tombstones LIMIT :max_places")?;
+
+        let mut result: HashMap<SyncGuid, OutgoingInfo> = HashMap::new();
+
+        // We want to limit to 5000 places - tombstones are arguably the
+        // most important, so we fetch these first.
+        let ts_rows = tombstones_stmt.query_map_named(
+                            &[(":max_places", &(max_places as u32))],
+                            |row| SyncGuid(row.get::<_, String>("guid")))?;
+        for r in ts_rows {
+            let guid = r?;
+            trace!("outgoing tombstone {:?}", &guid);
+            result.insert(guid.clone(), OutgoingInfo::Tombstone);
+        }
+
+        // Max records is now limited by how many tombstones we found.
+        let max_places_left = max_places - result.len();
+
+        // We write info about the records we are updating to a temp table.
+        // While we could carry this around in memory, we'll need a temp table
+        // in `finish_outgoing` anyway, because we execute a `NOT IN` query
+        // there - which, in a worst-case scenario, is a very large `NOT IN`
+        // set.
+        db.execute_all(&["DROP TABLE IF EXISTS temp_sync_updated_meta",
```

Comment:
Huh, interesting. Do you think it's worth creating this at startup the way places creates its temp tables (@linacambridge might have an idea of why to / not to do this?)

Reply:
That's a good question! I guess one argument is that we expect the majority of our users here to not have sync - but let's see what @lina says.

I've made all other requested changes and push them.

Reply:
Why not both? 😄  `CREATE TEMP TABLE IF NOT EXISTS temp_sync_updated_meta` here...